aesara/tensor/nlinalg.py

```diff
@@ -823,6 +824,87 @@ def tensorsolve(a, b, axes=None):
     return TensorSolve(axes)(a, b)
 
 
+class MatMul(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def make_node(self, a, b):
```

Comment:
As a means of confirming that the shapes are compatible with the operation, but in the context of partial static shape information (i.e. the old `TensorType.broadcastable`, and now the new `TensorType.shape`)?

If that's what you're asking, then&mdash;generally speaking&mdash;yeah, we consider every opportunity to do things like that, especially when it helps prevent more cryptic errors down the line.

If we're talking about a dot product operation, then we should be able to check that the last dimension of `a` has the same broadcast value as the first dimension of `b` (i.e. `a.type.broadcastable[-1] == b.type.broadcastable[0]`).  We could do even better now that `TensorType.shape` is available.

Reply:
```python
In [2]: a = np.random.rand(1, 2, 4, 5) 
In [3]: b = np.random.rand(1, 5, 7)
In [4]: np.matmul(a,b).shape
Out[4]: (1, 2, 4, 7)
In [9]: at.matmul(a, b).type.broadcastable
Out[9]: (False, False, False, False)
```
Just to be sure I understand @rlouf 's concern correctly, we're referring to the problem shown in the snippet above where the first element of the output's `broadcastable` is `False` when it should be `True` because of the `1`?. How would we know the shape inside of `make_node` in order to encode this information while creating the output node? I thought it was impossible?

Reply:
If I understood correctly, he wasn't necessarily talking about having the correct broadcast pattern for the `Type`s output by `MatMul` (although that's also very important), but that the broadcast patterns could be used for input validation in `MatMul.make_node`.

For example, take the following incompatible inputs:
```python
import numpy as np
import aesara.tensor as at


a = np.random.rand(1)
b = np.random.rand(2)

np.matmul(a, b)
# ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 1)
```

Our Aesara version of `np.matmul` should raise an error for the same inputs, as well, and it will; however, it will raise that error _at run-time_ and not when the symbolic graph is created.  Now, in this example, we might've been able to tell that the inputs were not compatible based on the partial shape information provided by the broadcast patterns:

```python
a_at = at.as_tensor(a)
b_at = at.as_tensor(b)

a_at.type.broadcastable
# (True,)
b_at.type.broadcastable
# (False,)
```
Unfortunately, this isn't really enough to work with, unless you take the broadcast patterns too strictly, because these boolean broadcast values really translate to static shape information like `a_at.type.shape == (1,)` and `b_at.type.shape == (None,)` (this was conceptually true even before #711 was merged).  The `None` indicates that we do not know the shape of `b_at` at "compile" time.  It could be _anything_, including `(1,)`, so we can't actually use it for input validation in this (and most) case(s).

Nevertheless, you seem to have found an example for which the broadcast pattern of the _output_ of `MatMul` isn't correct, and that should definitely be fixed.

Reply:
That's what I meant, sorry for not expressing myself more clearly.

Reply:
I implemented a mechanism to validate the input shapes at compile time. If overuse of conditionals makes the code a bit off-putting so any refactoring tips are welcome.